server/app/mutations/host_nodes/update.rb

```diff
@@ -18,7 +23,47 @@ def execute
 
       notify_grid(self.host_node.grid)
 
-      self.host_node
+      # Update availability if given and trigger appropriate actions
+      if self.availability && self.host_node.availability != self.availability
+
+        self.host_node.set(:availability => self.availability)
+
+        # Trigger needed action for stateful services, scheduling loop handles stateles services
+        case self.availability
+        when HostNode::Availability::ACTIVE
+          start_stateful_services(self.host_node)
+        when HostNode::Availability::DRAIN
+          stop_stateful_services(self.host_node)
+        end
+      end
+
+      self.host_node.reload
+    end
+
+    def stop_stateful_services(host_node)
+      host_node.grid_service_instances.each do |instance|
+        if instance.grid_service.stateful?
+          info "setting desired state to stopped for instance #{instance.grid_service.to_path}-#{instance.instance_number}"
+          instance.set(desired_state: 'stopped')
```

Comment:
I would have said something about a `kontena service deploy` of the drained stateful service setting the `GridServiceInstance` `desired_state` back to `running`, but that's moot, because the drained state filters out the node and prevents the stateful service from being deployed:

```
$ kontena service deploy redis-stateful
 [fail] Deploying service redis-stateful      
 [error] 500 : Cannot find applicable node for service instance development/null/redis-stateful-2: Strategy Scheduler::Strategy::HighAvailability did not find any node:
	⊛ Deployed instance development/null/redis-stateful-1 to node core-01
```